site/content/posts/2021-06-22-understanding-kubernetes-tests.md

```diff
@@ -0,0 +1,172 @@
+
+#### How to run the whole set of conformance tests
+
+```
+$ sonobuoy run --mode certified-conformance
+```
+
+#### How many tests are there
+
+~500 (varies by release)
+
+#### When to run certified-conformance mode
+
+- To obtain results that you submit to obtain Certified Kubernetes status
+- As a relatively thorough and standard check of cluster creation or configuration, typically in CI.
+
+#### Risks:
+
+- Be aware that the time to run the full set of conformance tests can run into hours, especially depending on the size of the cluster.
+- Some tests may be disruptive to existing workloads (more details below)
+
+## Disruptive tests
+
+**Disruptive tests** are all the tests that have the `[Disruptive]` tag. Most tests typically try to sandbox themselves: they create their own namespaces to run in and only use resources scoped to that namespace.  However, some Kubernetes features can’t be tested in this way, for instance, cluster level permissions or fault tolerance and recovery. By testing these features, there may be side-effects that impact other workloads in the cluster.
+
+As a result, it is not advisable to run these tests on clusters running production workloads.
+
+There isn’t any particular reason to run this set of tests on their own, but if you did want to run them, you would provide the tag to Sonobuoy’s e2e-focus flag just like any other tagged set of tests (described above).
+
+#### How many tests are there
+
+~50
+
+#### When to run them
+
+You would generally not want to just run the disruptive tests.
+Potentially if developing against Kubernetes itself or tooling related to failure recovery of clusters and workloads. Though there may be a more targeted set of tests relevant to your need.
+
+#### Risks
+
+As the name says, the tests are disruptive. Nodes may be taken down, API server made unresponsive, cluster permissions altered, or resources hogged.
+
+## Non-Disruptive Conformance Tests
+
+Now that we understand how tests are tagged and the meaning of conformance and disruptive tests, it should be clear that these non-disruptive conformance tests are just tests that have the `[Conformance]` tag and NOT the `[Disruptive]` tag.
+
+This group of tests is significant because the value of the conformance tests extends beyond the Certified Kubernetes Program. Since it tests the core of Kubernetes functionality, most users consider the conformance tests to be a good baseline to test the health of existing clusters so they will routinely run them to validate their cluster creation or to help catch and diagnose issues quickly. If they are running these tests on existing clusters, they should avoid running the disruptive tests or else they compromise the health of the cluster in question.
+
+Since this is the most common use case for Sonobuoy, it is the default choice for Sonobuoy.
+
+#### How to run them
+
+```
+$ sonobuoy run
+```
+
+This is the same thing as providing the tags directly (with a slight historical difference):
```

Comment:
What's that difference?

Reply:
So if you do `sonobuoy gen` you'll see that the skip is actually set to `\[Disruptive\]|NoExecuteTaintManager`. They are meant to target the same tests I believe. The point was that in "the long-ago time" we thought conformance tests wouldn't be disruptive, so we only had `focus=conformance` and no skip. Then I found some conformance tests that WERE disruptive, but not tagged as such. So I added the default skip to be `disruptive` (which would apply to 0 tests currently) and `NoExecuteTaintManager` (which hit a few tests which should all be marked disruptive but weren't at the time).

Now we probably could change it to just disruptive, but I havent spent the time to check.

Reply:
Oh sorry, I meant that it would be useful to explain what the difference is explicitly.

Reply:
Oh OK. I wasn't sure about that. I couldn't think of a brief way of mentioning they were different. So I felt like I either had to ignore the difference (in which case I didnt want to say they were the same when they werent exactly) or I had to fully explain (which seemed, as written above, too meaningless to mention).

Thoughts on which approacch is better or a different middle ground?

Reply:
Or you can just remove `(with a slight historical difference)` this bit. That would work.